### PR TITLE
[video] use ffmpeg av_display_rotation_get for rotation from metadata

### LINKED ISSUE #23762: Problem with Mobile Vertical Video .mp4 files when playing videos
Sorry my english is not so good.

Have a Problem with Mobile Vertical Video .mp4 files when playing videos via Windows 10 Pro ( PC ) and Android TV 11 ( Nvidia Shield TV ).

1. screenshot1 when playing the mobile vertical video via Kodi 21 Omega Alpha 3 ( 64 Bit ) = not OK.

Screenshot1

2. screenshot2 when playing the same vertical video via Kodi 20.2 Nexus ( 64 Bit ) = OK

Screenshot2

3. mobile vertical videos play normally and not in landscape mode via Kodi Nexus 20.2 ...Kodi 19.5 Matrix and Kodi 18.9 Leia as well as VLC.

4. I have attached this vertical video .mp4 file as a .zip file here to screenshot1+2 so that you can test it yourself...e.g. under Linux or another operating system.

[VID-20230318-WA0000.zip](https://github.com/xbmc/xbmc/files/12591495/VID-20230318-WA0000.zip)

5. Kodi Debug LOG

[kodi.zip](https://github.com/xbmc/xbmc/files/12591551/kodi.zip)

Translated with www.DeepL.com/Translator (free version)

### PR DESCRIPTION
**Pull request by enen92**
## Description
From ffmpeg changelog:

```
2017-03-27 - ddef3d902f - lavf 57.68.100 - avformat.h
  Deprecate that demuxers export the stream rotation angle in AVStream.metadata
  (via an entry named "rotate"). Use av_stream_get_side_data() with
  AV_PKT_DATA_DISPLAYMATRIX instead, and read the rotation angle with
  av_display_rotation_get(). The same is done for muxing. Instead of adding a
  "rotate" entry to AVStream.metadata, AV_PKT_DATA_DISPLAYMATRIX side data has
  to be added to the AVStream.
```

This fixes automatic rotation of videos (and thumbs) from the respective metadata.

## Motivation and context
Fix https://github.com/xbmc/xbmc/issues/23762

## How has this been tested?
Runtime tested on linux

## What is the effect on users?
Rotation should now work when set on video metadata

## Screenshots (if appropriate):

image
image

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp [hunk 1]
```diff
@@ -1721,9 +1722,17 @@ CDemuxStream* CDVDDemuxFFmpeg::AddStream(int streamIdx)
...
-          st->iOrientation = atoi(rtag->value);
+        uint8_t* displayMatrixSideData =
+            av_stream_get_side_data(pStream, AV_PKT_DATA_DISPLAYMATRIX, nullptr);
+        if (displayMatrixSideData)
+        {
+          const double tetha =
+              av_display_rotation_get(reinterpret_cast<int32_t*>(displayMatrixSideData));
+          if (!std::isnan(tetha))
+          {
+            st->iOrientation = ((static_cast<int>(-tetha) % 360) + 360) % 360;
```
Comment (fritsch) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:1733]: The angle (in degrees) by which the transformation rotates the frame counterclockwise. The angle will be in range [-180.0, 180.0], or NaN if the matrix is singular.

Comment (fritsch) [reply xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:1733]: Why do you need the %360?

Comment (fritsch) [reply xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:1733]: I think (int(-theta) + 360) %360 should be enough, right?

Comment (enen92) [reply xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:1733]: Got it from mpv but I agree `(int(-theta) + 360) %360 ` should suffix. Probably they divide first to ensure rounding?

Comment (fritsch) [reply xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp]:
```
    sd = av_frame_get_side_data(src, AV_FRAME_DATA_DISPLAYMATRIX);
    if (sd) {
        double r = av_display_rotation_get((int32_t *)(sd->data));
        if (!isnan(r))
            dst->params.rotate = (((int)(-r) % 360) + 360) % 360;
    }
 ```
 
 Yes, indeed. Following the second comment here: https://stackoverflow.com/questions/11720656/modulo-operation-with-negative-numbers I rather think this is the issue. Please keep what mpv had ... sorry for the noise.

Comment (enen92) [reply xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:1733]: I see. Thanks for looking into it - i've reverted to the previous version.
A green for merging is welcome :)

**Review by fritsch — APPROVED** [review]
Thanks!

**Comment by malvinas2** [comment]
Is it possible to create a backport for version 20/Nexus?

**Comment by enen92** [comment]
@malvinas2 can you handle the backport pull request? I don't really use nexus (on nightlies ever since)

**Comment by malvinas2** [comment]
@enen92
After I built Kodi 20.2 on my computer, I was surprised that the demo-video was playing with the correct rotation. 
However, as Ronny-nerd noted in his issue, the problem only related to Omega Alpha ("screenshot2 when playing the s**ame vertical video via Kodi 20.2 Nexus (64 Bit) = OK**").

Lesson learnt: Read the entire issue .....

**Comment by Pseudo98** [comment]
I have just tested this with 21.0 Beta 1 on a Fire TV 4K Max, with MediaCodec or MediaCodec (Surface) enabled it is still not correctly displaying orientation, with all hardware acceleration disabled size and orientation is correct.

**Comment by enen92** [comment]
@Pseudo98 that's unfortunately unrelated. It's a firmware bug, contact the manufacturer.

**Pull request merged**